File: libs/language/src/lib/merge/__tests__/DictionaryProcessor.test.ts
import { describe, it, expect } from 'vitest';
import { BlueNode } from '../../model';
import { Merger } from '../Merger';
import { MergingProcessor } from '../MergingProcessor';
import { SequentialMergingProcessor } from '../processors/SequentialMergingProcessor';
import { DictionaryProcessor } from '../processors/DictionaryProcessor';
import { TypeAssigner } from '../processors/TypeAssigner';
import { NodeExtender } from '../../utils/NodeExtender';
import { NO_LIMITS } from '../../utils/limits';
import {
  DICTIONARY_TYPE,
  TEXT_TYPE,
  CORE_TYPE_BLUE_ID_TO_NAME_MAP,
} from '../../utils/Properties';
import { BasicNodeProvider } from '../../provider/BasicNodeProvider';

describe('DictionaryProcessor', () => {
  it('testKeyTypeAndValueTypeAssignment', () => {
    const dictA = new BlueNode()
      .setName('DictA')
      .setType('Dictionary')
      .setKeyType('Text')
      .setValueType('Integer');

    const nodeProvider = new BasicNodeProvider();
    nodeProvider.addSingleNodes(dictA);

    const dictB = new BlueNode()
      .setName('DictB')
      .setType(new BlueNode().setBlueId(nodeProvider.getBlueIdByName('DictA')));

    nodeProvider.addSingleNodes(dictB);
    const mergingProcessor: MergingProcessor = new SequentialMergingProcessor([
      new DictionaryProcessor(),
    ]);

    const merger = new Merger(mergingProcessor, nodeProvider);
    const dictANode = nodeProvider.findNodeByName('DictA');
    if (!dictANode) {
      throw new Error('DictA not found');
    }
    const result = merger.resolve(dictANode, NO_LIMITS);

    expect(result.getKeyType()?.getBlueId()).toBeDefined();
    expect(result.getValueType()?.getBlueId()).toBeDefined();
    expect(
      CORE_TYPE_BLUE_ID_TO_NAME_MAP[
        result
          .getKeyType()!
          .getBlueId() as keyof typeof CORE_TYPE_BLUE_ID_TO_NAME_MAP
      ]
    ).toBe('Text');
    expect(
      CORE_TYPE_BLUE_ID_TO_NAME_MAP[
        result
          .getValueType()!
          .getBlueId() as keyof typeof CORE_TYPE_BLUE_ID_TO_NAME_MAP
      ]
    ).toBe('Integer');
  });

  it('testDictionaryWithValidTypes', () => {
    const nodeProvider = new BasicNodeProvider();

    const a = 'name: A';
    nodeProvider.addSingleDocs(a);

    const b = `name: B
type:
  blueId: ${nodeProvider.getBlueIdByName('A')}`;
    nodeProvider.addSingleDocs(b);

    const dictOfAToB = `name: DictOfAToB
type: Dictionary
keyType: ${TEXT_TYPE}
valueType:
  blueId: ${nodeProvider.getBlueIdByName('A')}
key1:
  type:
    blueId: ${nodeProvider.getBlueIdByName('A')}
key2:
  type:
    blueId: ${nodeProvider.getBlueIdByName('B')}`;
    nodeProvider.addSingleDocs(dictOfAToB);

    const mergingProcessor: MergingProcessor = new SequentialMergingProcessor([
      new TypeAssigner(),
      new DictionaryProcessor(),
    ]);

    const merger = new Merger(mergingProcessor, nodeProvider);
    const dictOfAToBNode = nodeProvider.findNodeByName('DictOfAToB');
    if (!dictOfAToBNode) {
      throw new Error('DictOfAToB not found');
    }
    new NodeExtender(nodeProvider).extend(dictOfAToBNode, NO_LIMITS);
    const result = merger.resolve(dictOfAToBNode, NO_LIMITS);

    expect(result.getKeyType()?.getBlueId()).toBeDefined();
    expect(
      CORE_TYPE_BLUE_ID_TO_NAME_MAP[
        result
          .getKeyType()!
          .getBlueId() as keyof typeof CORE_TYPE_BLUE_ID_TO_NAME_MAP
      ]
    ).toBe('Text');
    expect(result.getValueType()?.getName()).toBe('A');
    expect(Object.keys(result.getProperties() || {}).length).toBe(2);
    expect(result.getProperties()?.['key1']?.getType()?.getName()).toBe('A');
    expect(result.getProperties()?.['key2']?.getType()?.getName()).toBe('B');
  });

  it('testDictionaryWithInvalidKeyType', () => {
    const nodeProvider = new BasicNodeProvider();

    const dictWithInvalidKeyType = `name: DictWithInvalidKeyType
type: Dictionary
keyType: ${DICTIONARY_TYPE}
valueType: ${TEXT_TYPE}`;
    nodeProvider.addSingleDocs(dictWithInvalidKeyType);

    const mergingProcessor: MergingProcessor = new SequentialMergingProcessor([
      new TypeAssigner(),
      new DictionaryProcessor(),
    ]);

    const merger = new Merger(mergingProcessor, nodeProvider);
    const dictNode = nodeProvider.findNodeByName('DictWithInvalidKeyType');
    if (!dictNode) {
      throw new Error('DictWithInvalidKeyType not found');
    }
    new NodeExtender(nodeProvider).extend(dictNode, NO_LIMITS);

    expect(() => merger.resolve(dictNode, NO_LIMITS)).toThrow();
  });

  it('testDictionaryWithInvalidValueType', () => {
    const nodeProvider = new BasicNodeProvider();

    const a = 'name: A';
    nodeProvider.addSingleDocs(a);

    const dictWithInvalidValue = `name: DictWithInvalidValue
type: Dictionary
keyType: ${TEXT_TYPE}
valueType:
  blueId: ${nodeProvider.getBlueIdByName('A')}
key1:
  type: ${TEXT_TYPE}`; // This should cause an error
    nodeProvider.addSingleDocs(dictWithInvalidValue);

    const mergingProcessor: MergingProcessor = new SequentialMergingProcessor([
      new TypeAssigner(),
      new DictionaryProcessor(),
    ]);

    const merger = new Merger(mergingProcessor, nodeProvider);
    const dictNode = nodeProvider.findNodeByName('DictWithInvalidValue');
    if (!dictNode) {
      throw new Error('DictWithInvalidValue not found');
    }
    new NodeExtender(nodeProvider).extend(dictNode, NO_LIMITS);

    expect(() => merger.resolve(dictNode, NO_LIMITS)).toThrow();
  });

  it('testNonDictionaryTypeWithKeyTypeOrValueType', () => {
    const nodeProvider = new BasicNodeProvider();

    const nonDictWithKeyType = `name: NonDictWithKeyType
type: ${TEXT_TYPE}
keyType: ${TEXT_TYPE}`;
    nodeProvider.addSingleDocs(nonDictWithKeyType);

    const mergingProcessor: MergingProcessor = new SequentialMergingProcessor([
      new TypeAssigner(),
      new DictionaryProcessor(),
    ]);

    const merger = new Merger(mergingProcessor, nodeProvider);
    const nonDictNode = nodeProvider.findNodeByName('NonDictWithKeyType');
    if (!nonDictNode) {
      throw new Error('NonDictWithKeyType not found');
    }
    new NodeExtender(nodeProvider).extend(nonDictNode, NO_LIMITS);

    expect(() => merger.resolve(nonDictNode, NO_LIMITS)).toThrow();
  });
});

File: references/java-sdk/src/test/java/blue/language/DictionaryProcessorTest.java
package blue.language;

import blue.language.merge.Merger;
import blue.language.merge.MergingProcessor;
import blue.language.model.Node;
import blue.language.merge.processor.DictionaryProcessor;
import blue.language.merge.processor.SequentialMergingProcessor;
import blue.language.merge.processor.TypeAssigner;
import blue.language.provider.BasicNodeProvider;
import blue.language.utils.NodeExtender;
import blue.language.utils.limits.Limits;
import org.junit.jupiter.api.Test;

import java.util.Arrays;

import static blue.language.blueid.legacy.LegacyBlueIdCalculator.calculateBlueId;
import static blue.language.utils.Properties.*;
import static org.junit.jupiter.api.Assertions.*;

public class DictionaryProcessorTest {

    @Test
    public void testKeyTypeAndValueTypeAssignment() {
        Node dictA = new Node().name("DictA")
                .type("Dictionary")
                .keyType("Text")
                .valueType("Integer");
        Node dictB = new Node().name("DictB")
                .type(new Node().blueId(calculateBlueId(dictA)));

        BasicNodeProvider nodeProvider = new BasicNodeProvider(Arrays.asList(dictA, dictB));
        MergingProcessor mergingProcessor = new SequentialMergingProcessor(
                Arrays.asList(
                        new DictionaryProcessor()
                )
        );

        Merger merger = new Merger(mergingProcessor, nodeProvider);
        Node dictANode = nodeProvider.findNodeByName("DictA").orElseThrow(() -> new IllegalStateException("No \"DictA\" available for NodeProvider."));
        Node result = merger.resolve(dictANode, Limits.NO_LIMITS);

        assertEquals("Text", CORE_TYPE_BLUE_ID_TO_NAME_MAP.get(result.getKeyType().getBlueId()));
        assertEquals("Integer", CORE_TYPE_BLUE_ID_TO_NAME_MAP.get(result.getValueType().getBlueId()));
    }

    @Test
    public void testDictionaryWithValidTypes() throws Exception {
        BasicNodeProvider nodeProvider = new BasicNodeProvider();

        String a = "name: A";
        nodeProvider.addSingleDocs(a);

        String b = "name: B\n" +
                   "type:\n" +
                   "  blueId: " + nodeProvider.getBlueIdByName("A");
        nodeProvider.addSingleDocs(b);

        String dictOfAToB = "name: DictOfAToB\n" +
                            "type: Dictionary\n" +
                            "keyType: " + TEXT_TYPE + "\n" +
                            "valueType: \n" +
                            "  blueId: " + nodeProvider.getBlueIdByName("A") + "\n" +
                            "key1:\n" +
                            "  type:\n" +
                            "    blueId: " + nodeProvider.getBlueIdByName("A") + "\n" +
                            "key2:\n" +
                            "  type:\n" +
                            "    blueId: " + nodeProvider.getBlueIdByName("B");
        nodeProvider.addSingleDocs(dictOfAToB);

        MergingProcessor mergingProcessor = new SequentialMergingProcessor(
                Arrays.asList(
                        new TypeAssigner(),
                        new DictionaryProcessor()
                )
        );

        Merger merger = new Merger(mergingProcessor, nodeProvider);
        Node dictOfAToBNode = nodeProvider.getNodeByName("DictOfAToB");
        new NodeExtender(nodeProvider).extend(dictOfAToBNode, Limits.NO_LIMITS);
        Node result = merger.resolve(dictOfAToBNode);

        assertEquals("Text", CORE_TYPE_BLUE_ID_TO_NAME_MAP.get(result.getKeyType().getBlueId()));
        assertEquals("A", result.getValueType().getName());
        assertEquals(2, result.getProperties().size());
        assertEquals("A", result.getProperties().get("key1").getType().getName());
        assertEquals("B", result.getProperties().get("key2").getType().getName());
    }

    @Test
    public void testDictionaryWithInvalidKeyType() throws Exception {
        BasicNodeProvider nodeProvider = new BasicNodeProvider();

        String dictWithInvalidKeyType = "name: DictWithInvalidKeyType\n" +
                                        "type: Dictionary\n" +
                                        "keyType: " + DICTIONARY_TYPE + "\n" +
                                        "valueType: " + TEXT_TYPE;
        nodeProvider.addSingleDocs(dictWithInvalidKeyType);

        MergingProcessor mergingProcessor = new SequentialMergingProcessor(
                Arrays.asList(
                        new TypeAssigner(),
                        new DictionaryProcessor()
                )
        );

        Merger merger = new Merger(mergingProcessor, nodeProvider);
        Node dictNode = nodeProvider.findNodeByName("DictWithInvalidKeyType").orElseThrow(() -> new IllegalStateException("No \"DictWithInvalidKeyType\" available for NodeProvider."));
        new NodeExtender(nodeProvider).extend(dictNode, Limits.NO_LIMITS);

        assertThrows(IllegalArgumentException.class, () -> merger.resolve(dictNode));
    }

    @Test
    public void testDictionaryWithInvalidValueType() throws Exception {
        BasicNodeProvider nodeProvider = new BasicNodeProvider();

        String a = "name: A";
        nodeProvider.addSingleDocs(a);

        String dictWithInvalidValue = "name: DictWithInvalidValue\n" +
                                      "type: Dictionary\n" +
                                      "keyType: " + TEXT_TYPE + "\n" +
                                      "valueType: \n" +
                                      "  blueId: " + nodeProvider.getBlueIdByName("A") + "\n" +
                                      "key1:\n" +
                                      "  type: " + TEXT_TYPE;  // This should cause an error
        nodeProvider.addSingleDocs(dictWithInvalidValue);

        MergingProcessor mergingProcessor = new SequentialMergingProcessor(
                Arrays.asList(
                        new TypeAssigner(),
                        new DictionaryProcessor()
                )
        );

        Merger merger = new Merger(mergingProcessor, nodeProvider);
        Node dictNode = nodeProvider.findNodeByName("DictWithInvalidValue").orElseThrow(() -> new IllegalStateException("No \"DictWithInvalidValue\" available for NodeProvider."));
        new NodeExtender(nodeProvider).extend(dictNode, Limits.NO_LIMITS);

        assertThrows(IllegalArgumentException.class, () -> merger.resolve(dictNode));
    }

    @Test
    public void testNonDictionaryTypeWithKeyTypeOrValueType() throws Exception {
        BasicNodeProvider nodeProvider = new BasicNodeProvider();

        String nonDictWithKeyType = "name: NonDictWithKeyType\n" +
                                    "type: " + TEXT_TYPE + "\n" +
                                    "keyType: " + TEXT_TYPE;
        nodeProvider.addSingleDocs(nonDictWithKeyType);

        MergingProcessor mergingProcessor = new SequentialMergingProcessor(
                Arrays.asList(
                        new TypeAssigner(),
                        new DictionaryProcessor()
                )
        );

        Merger merger = new Merger(mergingProcessor, nodeProvider);
        Node nonDictNode = nodeProvider.findNodeByName("NonDictWithKeyType").orElseThrow(() -> new IllegalStateException("No \"NonDictWithKeyType\" available for NodeProvider."));
        new NodeExtender(nodeProvider).extend(nonDictNode, Limits.NO_LIMITS);

        assertThrows(IllegalArgumentException.class, () -> merger.resolve(nonDictNode));
    }

}